homework-selects/sql_questions.sql

```diff
@@ -3,55 +3,460 @@
+SELECT employee_id, (first_name || ' ' || last_name) AS employee_name
+FROM employees
+WHERE department_id in (
+    SELECT department_id
+    FROM employees
+    WHERE (first_name || ' ' || last_name) LIKE '%T%'
+);
+
 --25. full name(first and last name), job title, starting and ending date of last jobs for those employees with worked without a commission percentage.
+SELECT (first_name || ' ' || last_name) AS employee_name, job_title, last_start_date, last_end_date
+FROM employees
+    INNER JOIN (
+        SELECT employee_id, MAX (start_date) as last_start_date, MAX (end_date) as last_end_date
+        FROM job_history
+        GROUP BY employee_id
+    ) USING (employee_id)
+    INNER JOIN jobs USING (job_id)
+WHERE commission_pct IS NULL;
+
 --26. the employee number, name( first name and last name ), and salary for all employees who earn more than the average salary and who work in a department with any employee with a J in their name.
+SELECT employee_id, (first_name || ' ' || last_name) AS employee_name, salary
+FROM employees
+WHERE salary > (
+    SELECT AVG(salary)
+    FROM employees
+)
+  AND department_id in (
+    SELECT department_id
+    FROM employees
+    WHERE (first_name || ' ' || last_name) LIKE '%J%'
+);
+
 --27. the employee number, name( first name and last name ) and job title for all employees whose salary is smaller than any salary of those employees whose job title is MK_MAN.
+SELECT employee_id, (first_name || ' ' || last_name) AS employee_name, job_title
+FROM employees
+    INNER JOIN jobs USING (job_id)
+WHERE salary < (
+    SELECT MIN(salary)
+    FROM employees
+    WHERE job_id = 'MK_MAN'
+);
+
 --28. the employee number, name( first name and last name ) and job title for all employees whose salary is smaller than any salary of those employees whose job title is MK_MAN. Exclude Job title MK_MAN.
+SELECT employee_id, (first_name || ' ' || last_name) AS employee_name, job_title
+FROM employees
+    INNER JOIN jobs USING (job_id)
+WHERE salary < (
+    SELECT MIN(salary)
+    FROM employees
+    WHERE job_id = 'MK_MAN'
+) AND job_id != 'MK_MAN'; --Redundant: at this point any salary is smaller than MIN(salary) with job_id 'MK_MAN', thus, job_id will never be 'MK_MAN'
+
+
 --29. all the information of those employees who did not have any job in the past.
+SELECT *
+FROM employees
+WHERE employee_id NOT IN (
+    SELECT DISTINCT EMPLOYEE_ID
+    FROM job_history
+    );
+
 --30. the employee number, name( first name and last name ) and job title for all employees whose salary is more than any average salary of any department.
+SELECT employee_id, (first_name || ' ' || last_name) AS employee_name, job_title
+FROM employees
+    INNER JOIN jobs USING (job_id)
+WHERE salary > (
+    SELECT MIN(average_salaries_table.average_salary)
+    FROM (
+         SELECT AVG(salary) AS average_salary
+         FROM EMPLOYEES
+         ) average_salaries_table
+    );
```

Comment:
Hm, can we refactor this to use JOIN instead of SUB-QUERY? To have smth like: 
```suggestion
WHERE salary > > ANY (SELECT AVG(salary)
                        FROM departments
                                 LEFT JOIN employees USING (department_id)
                        GROUP BY department_id);
```

Reply:
Refactored to use **JOIN**